source/common/tracing/http_tracer_impl.cc

```diff
@@ -84,10 +84,40 @@ static void addGrpcRequestTags(Span& span, const Http::RequestHeaderMap& headers
 template <class T> static void addGrpcResponseTags(Span& span, const T& headers) {
   addTagIfNotNull(span, Tracing::Tags::get().GrpcStatusCode, headers.GrpcStatus());
   addTagIfNotNull(span, Tracing::Tags::get().GrpcMessage, headers.GrpcMessage());
-  // Set error tag when status is not OK.
+  // Set error tag when Grpc status code represents an upstream error. See
+  // https://github.com/envoyproxy/envoy/issues/18877.
   absl::optional<Grpc::Status::GrpcStatus> grpc_status_code = Grpc::Common::getGrpcStatus(headers);
-  if (grpc_status_code && grpc_status_code.value() != Grpc::Status::WellKnownGrpcStatus::Ok) {
-    span.setTag(Tracing::Tags::get().Error, Tracing::Tags::get().True);
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.update_grpc_response_error_tag")) {
+    if (grpc_status_code.has_value()) {
+      const auto& status = grpc_status_code.value();
+      if (status != Grpc::Status::WellKnownGrpcStatus::InvalidCode) {
+        switch (status) {
+        // Each case below is considered to be a client side error, therefore should not be
+        // tagged as an upstream error. See https://grpc.github.io/grpc/core/md_doc_statuscodes.html
+        // for more details about how each Grpc status code is defined and whether it is an
+        // upstream error or a client error.
+        case Grpc::Status::WellKnownGrpcStatus::Ok:
```

Comment:
the CI failure is because most of the new lines are not tested.  it should be pretty easy to add unit tests the values are what we expect.
/wait

Reply:
sorry this still has a default, where now if we get an unknown code we panic if we receive it.  Can we not remove the default entirely?
/wait

Reply:
IIUC, the default should be removed right?

Reply:
So it's always possible to get an out of bounds gRPC status, which we should handle as gracefully as we can (say ENVOY_BUG and treat as an error).  But if we remove the default from the switch statement and we import a version of gRPC with WellKnownGrpcStatus::NewStatus the import won't even compile unless we add to the switch statement here.  That's what I'm asking for :-)

Reply:
Oh, I see, that makes sense. Thank you @alyssawilk for explanation!